pyoozie/tags.py

```diff
@@ -494,6 +496,133 @@ def _xml(self, doc, tag, text):
         return doc
 
 
+class _WorkflowEntity(typing.Iterable):
+    """An abstract object representing an Oozie workflow action that can be serialized to XML."""
+    # pylint: disable=abstract-method
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(
+            self,
+            xml_tag,       # typing.Text
+            name=None,     # type: typing.Optional[typing.Text]
+            on_error=None  # type: typing.Optional[_WorkflowEntity]
+    ):
+        # type: (...) -> None
+        self.xml_tag = xml_tag
+        self.__name = name
+        self.__on_error = copy.deepcopy(on_error)
+        self.__uid = uuid.uuid4().hex[:8]
+        self.__identifier = self.create_identifier(xml_tag)
+
+    def create_identifier(self, xml_tag):
+        # type: (typing.Text) -> typing.Text
+        identifier = None
+        if self.__name:
+            identifier = '{tag}-{name}'.format(tag=xml_tag, name=self.__name)
+        else:
+            identifier = '{tag}-{uid}'.format(tag=xml_tag, uid=self.__uid)
+        return validate_xml_id(identifier)
+
+    def identifier(self):
+        # type: () -> typing.Text
+        return self.__identifier
+
+    def _xml_and_get_on_error(self, doc, tag, text, on_next, on_error):
+        # type: (yattag.doc.Doc, yattag.doc.Doc.tag, yattag.doc.Doc.text, typing.Text, typing.Text) -> yattag.doc.Doc
+        if self.__on_error:
+            self.__on_error._xml(doc, tag, text, on_next, on_error)
+        return self.__on_error.identifier() if self.__on_error else (
+            on_error if on_error else on_next
+        )
+
+    @abc.abstractmethod
+    def _xml(self, doc, tag, text, on_next, on_error):
+        # type: (yattag.doc.Doc, yattag.doc.Doc.tag, yattag.doc.Doc.text, typing.Text, typing.Text) -> yattag.doc.Doc
+        raise NotImplementedError()
+
+    def __iter__(self):
+        # type: () -> typing.Generator[_WorkflowEntity, None, None]
+        yield self
+        if self.__on_error:
+            for action in self.__on_error:
+                yield action
+
+    def __bool__(self):  # type: () -> bool
+        return bool(list(self))
+
+    def __nonzero__(self):  # type: () -> bool
+        return self.__bool__()
+
+    def __repr__(self):  # type: () -> str
+        return str('{_class}({identifier})'.format(_class=type(self).__name__, identifier=self.identifier()))
```

Comment:
This is nitpicky, but could you please surround the return value with `<` and `>`? The string you're building here looks like something you could paste into a REPL to get back an equivalent object, but I'm guessing it doesn't actually round-trip due to the mismatch between the constructor parameter `xml_tag` and the `identifier` you're using here.

Motivation:

> For many types, this function makes an attempt to return a string that would yield an object with the same value when passed to eval(), otherwise the representation is a string enclosed in angle brackets that contains the name of the type of the object together with additional information often including the name and address of the object.

You could of course create a representation that _does_ round-trip, but that would make your identifier generation code a little more complicated.